File: backend/routes/queue.py
"""Queue routes for the mt music player API."""

from backend.services.database import DatabaseService, get_db
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

router = APIRouter(prefix="/queue", tags=["queue"])


class QueueAddRequest(BaseModel):
    """Request to add tracks to the queue."""

    track_ids: list[int] = Field(min_length=1)
    position: int | None = None


class QueueAddFilesRequest(BaseModel):
    """Request to add files directly to the queue."""

    filepaths: list[str] = Field(min_length=1)
    position: int | None = None


class QueueReorderRequest(BaseModel):
    """Request to reorder tracks in the queue."""

    from_position: int = Field(ge=0)
    to_position: int = Field(ge=0)


class QueueShuffleRequest(BaseModel):
    """Request to shuffle the queue."""

    keep_current: bool = True


@router.get("")
async def get_queue(db: DatabaseService = Depends(get_db)):
    """Get the current playback queue."""
    items = db.get_queue()
    return {
        "items": items,
        "count": len(items),
    }


@router.post("/add", status_code=201)
async def add_to_queue(request: QueueAddRequest, db: DatabaseService = Depends(get_db)):
    """Add track(s) to the queue."""
    added = db.add_to_queue(request.track_ids, request.position)
    queue_length = db.get_queue_length()
    return {
        "added": added,
        "queue_length": queue_length,
    }


@router.post("/add-files", status_code=201)
async def add_files_to_queue(request: QueueAddFilesRequest, db: DatabaseService = Depends(get_db)):
    """Add files directly to the queue (for drag-and-drop)."""
    added, tracks = db.add_files_to_queue(request.filepaths, request.position)
    queue_length = db.get_queue_length()
    return {
        "added": added,
        "queue_length": queue_length,
        "tracks": tracks,
    }


@router.delete("/{position}", status_code=204)
async def remove_from_queue(position: int, db: DatabaseService = Depends(get_db)):
    """Remove a track from the queue by position."""
    if not db.remove_from_queue(position):
        raise HTTPException(status_code=404, detail=f"No track at position {position}")


@router.post("/clear", status_code=204)
async def clear_queue(db: DatabaseService = Depends(get_db)):
    """Clear the entire queue."""
    db.clear_queue()


@router.post("/reorder")
async def reorder_queue(request: QueueReorderRequest, db: DatabaseService = Depends(get_db)):
    """Reorder tracks in the queue."""
    if not db.reorder_queue(request.from_position, request.to_position):
        raise HTTPException(status_code=400, detail="Invalid positions")
    return {
        "success": True,
        "queue_length": db.get_queue_length(),
    }


@router.post("/shuffle")
async def shuffle_queue(request: QueueShuffleRequest, db: DatabaseService = Depends(get_db)):
    """Shuffle the queue."""
    import random

    items = db.get_queue()
    if not items:
        return {"success": True, "queue_length": 0}

    # Get filepaths
    filepaths = [item["track"]["filepath"] for item in items]

    if request.keep_current and filepaths:
        # Keep first item, shuffle rest
        first = filepaths[0]
        rest = filepaths[1:]
        random.shuffle(rest)
        filepaths = [first] + rest
    else:
        random.shuffle(filepaths)

    # Rebuild queue
    db.clear_queue()
    for filepath in filepaths:
        # Get track ID for filepath
        track = db.get_track_by_filepath(filepath)
        if track:
            db.add_to_queue([track["id"]])

    return {
        "success": True,
        "queue_length": db.get_queue_length(),
    }

File: src-tauri/src/commands/queue.rs
//! Tauri commands for queue management.
//!
//! These commands expose queue operations to the frontend,
//! replacing the Python FastAPI queue routes.

use rand::rng;
use rand::seq::SliceRandom;
use tauri::{AppHandle, Emitter, State};

use crate::db::{queue, Database, QueueItem, Track};

/// Response for queue get operations
#[derive(Clone, serde::Serialize)]
pub struct QueueResponse {
    pub items: Vec<QueueItem>,
    pub count: i64,
}

/// Response for queue add operations
#[derive(Clone, serde::Serialize)]
pub struct QueueAddResponse {
    pub added: i64,
    pub queue_length: i64,
}

/// Response for queue add-files operations
#[derive(Clone, serde::Serialize)]
pub struct QueueAddFilesResponse {
    pub added: i64,
    pub queue_length: i64,
    pub tracks: Vec<Track>,
}

/// Response for queue operations that return success status
#[derive(Clone, serde::Serialize)]
pub struct QueueOperationResponse {
    pub success: bool,
    pub queue_length: i64,
}

/// Get the current playback queue with track metadata
#[tauri::command]
pub fn queue_get(db: State<'_, Database>) -> Result<QueueResponse, String> {
    let conn = db.conn().map_err(|e| e.to_string())?;
    let items = queue::get_queue(&conn).map_err(|e| e.to_string())?;
    let count = items.len() as i64;

    Ok(QueueResponse { items, count })
}

/// Add tracks to the queue by track IDs
#[tauri::command]
pub fn queue_add(
    app: AppHandle,
    db: State<'_, Database>,
    track_ids: Vec<i64>,
    position: Option<i64>,
) -> Result<QueueAddResponse, String> {
    if track_ids.is_empty() {
        return Err("track_ids must not be empty".to_string());
    }

    let conn = db.conn().map_err(|e| e.to_string())?;
    let added = queue::add_to_queue(&conn, &track_ids, position).map_err(|e| e.to_string())?;
    let queue_length = queue::get_queue_length(&conn).map_err(|e| e.to_string())?;

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(QueueAddResponse {
        added,
        queue_length,
    })
}

/// Add files directly to the queue (for drag-and-drop support)
#[tauri::command]
pub fn queue_add_files(
    app: AppHandle,
    db: State<'_, Database>,
    filepaths: Vec<String>,
    position: Option<i64>,
) -> Result<QueueAddFilesResponse, String> {
    if filepaths.is_empty() {
        return Err("filepaths must not be empty".to_string());
    }

    let conn = db.conn().map_err(|e| e.to_string())?;
    let (added, tracks) =
        queue::add_files_to_queue(&conn, &filepaths, position).map_err(|e| e.to_string())?;
    let queue_length = queue::get_queue_length(&conn).map_err(|e| e.to_string())?;

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(QueueAddFilesResponse {
        added,
        queue_length,
        tracks,
    })
}

/// Remove a track from the queue by position
#[tauri::command]
pub fn queue_remove(
    app: AppHandle,
    db: State<'_, Database>,
    position: i64,
) -> Result<(), String> {
    let conn = db.conn().map_err(|e| e.to_string())?;
    let removed = queue::remove_from_queue(&conn, position).map_err(|e| e.to_string())?;

    if !removed {
        return Err(format!("No track at position {}", position));
    }

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(())
}

/// Clear the entire queue
#[tauri::command]
pub fn queue_clear(app: AppHandle, db: State<'_, Database>) -> Result<(), String> {
    let conn = db.conn().map_err(|e| e.to_string())?;
    queue::clear_queue(&conn).map_err(|e| e.to_string())?;

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(())
}

/// Reorder tracks in the queue (move from one position to another)
#[tauri::command]
pub fn queue_reorder(
    app: AppHandle,
    db: State<'_, Database>,
    from_position: i64,
    to_position: i64,
) -> Result<QueueOperationResponse, String> {
    let conn = db.conn().map_err(|e| e.to_string())?;
    let success =
        queue::reorder_queue(&conn, from_position, to_position).map_err(|e| e.to_string())?;

    if !success {
        return Err("Invalid positions".to_string());
    }

    let queue_length = queue::get_queue_length(&conn).map_err(|e| e.to_string())?;

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(QueueOperationResponse {
        success,
        queue_length,
    })
}

/// Shuffle the queue using Fisher-Yates algorithm
#[tauri::command]
pub fn queue_shuffle(
    app: AppHandle,
    db: State<'_, Database>,
    keep_current: Option<bool>,
) -> Result<QueueOperationResponse, String> {
    let conn = db.conn().map_err(|e| e.to_string())?;
    let items = queue::get_queue(&conn).map_err(|e| e.to_string())?;

    if items.is_empty() {
        return Ok(QueueOperationResponse {
            success: true,
            queue_length: 0,
        });
    }

    // Get filepaths from queue items
    let mut filepaths: Vec<String> = items.iter().map(|item| item.track.filepath.clone()).collect();

    let keep_current = keep_current.unwrap_or(true);

    if keep_current && !filepaths.is_empty() {
        // Keep first item, shuffle rest using Fisher-Yates
        let first = filepaths.remove(0);
        filepaths.shuffle(&mut rng());
        filepaths.insert(0, first);
    } else {
        // Shuffle all items
        filepaths.shuffle(&mut rng());
    }

    // Rebuild queue with shuffled order
    queue::clear_queue(&conn).map_err(|e| e.to_string())?;

    for filepath in &filepaths {
        conn.execute(
            "INSERT INTO queue (filepath) VALUES (?)",
            rusqlite::params![filepath],
        )
        .map_err(|e| e.to_string())?;
    }

    let queue_length = queue::get_queue_length(&conn).map_err(|e| e.to_string())?;

    // Emit queue updated event
    let _ = app.emit("queue:updated", ());

    Ok(QueueOperationResponse {
        success: true,
        queue_length,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_queue_response_serialization() {
        let response = QueueResponse {
            items: vec![],
            count: 0,
        };

        let json = serde_json::to_string(&response).unwrap();
        assert!(json.contains("\"count\":0"));
        assert!(json.contains("\"items\":[]"));
    }

    #[test]
    fn test_queue_add_response_serialization() {
        let response = QueueAddResponse {
            added: 3,
            queue_length: 10,
        };

        let json = serde_json::to_string(&response).unwrap();
        assert!(json.contains("\"added\":3"));
        assert!(json.contains("\"queue_length\":10"));
    }
}
